### PR TITLE
Moving workflow userstore component to user-mgt and handling user store operations

### PR DESCRIPTION
**Pull request by cdwijayarathna**

### CODE REVIEW

**Comment by cdwijayarathna** [comment]
Should merge https://github.com/wso2/product-is/pull/166 with this.

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java [hunk 1]
```diff
@@ -84,13 +92,33 @@ public boolean startAddUserFlow(String userStoreDomain, String userName, Object
...
+        String uuid = UUID.randomUUID().toString();
+        boolean state = startWorkFlow(wfParams, nonWfParams, uuid);
+
+        //WF_REQUEST_ENTITY_RELATIONSHIP table has foreign key to WF_REQUEST, so need to run this after WF_REQUEST is
+        // updated
+        if (!Boolean.TRUE.equals(getWorkFlowCompleted()) && !state) {
+            //ToDo: Add thread local to handle scenarios where workflow is not associated with the event.
+            try {
+                workflowService.addRequestEntityRelationships(uuid, new Entity[]{new Entity(fullyQualifiedName,
+                        "USER")});
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java]: Replace this with a constant

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java [hunk 2]
```diff
@@ -84,13 +92,33 @@ public boolean startAddUserFlow(String userStoreDomain, String userName, Object
...
+
+            } catch (InternalWorkflowException e) {
+                //Ignore exception which occurs at DB level since no workflows associated with event
+                log.info("No workflow associated with the operation.");
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java]: Why is this exception not being logged?

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/DeleteClaimWFRequestHandler.java [hunk 3]
```diff
@@ -57,13 +63,35 @@
...
-        return startWorkFlow(wfParams, nonWfParams);
+        String uuid = UUID.randomUUID().toString();
+        boolean state = startWorkFlow(wfParams, nonWfParams, uuid);
+
+        //WF_REQUEST_ENTITY_RELATIONSHIP table has foreign key to WF_REQUEST, so need to run this after WF_REQUEST is
+        // updated
+        if (!Boolean.TRUE.equals(getWorkFlowCompleted()) && !state) {
+            try {
+                workflowService.addRequestEntityRelationships(uuid, new Entity[]{new Entity(fullyQualifiedName,
+                        "USER"), new Entity(claimURI, "CLAIM")});
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/DeleteClaimWFRequestHandler.java]: Replace the hard-coded strings with constants

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/DeleteRoleWFRequestHandler.java [hunk 4]
```diff
@@ -53,11 +59,35 @@
...
+        // updated
+        if (!Boolean.TRUE.equals(getWorkFlowCompleted()) && !state) {
+            //ToDo: Add thread local to handle scenarios where workflow is not associated with the event.
+            try {
+                workflowService.addRequestEntityRelationships(uuid, new Entity[]{new Entity(fullyQualifiedName,
+                        "ROLE")});
+
+            } catch (InternalWorkflowException e) {
+                //Ignore exception which occurs at DB level since no workflows associated with event
+                log.info("No workflow associated with the operation.");
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/DeleteRoleWFRequestHandler.java]: Exception is not logged or thrown

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/DeleteUserWFRequestHandler.java [hunk 5]
```diff
@@ -53,11 +59,34 @@
...
+        // updated
+        if (!Boolean.TRUE.equals(getWorkFlowCompleted()) && !state) {
+            //ToDo: Add thread local to handle scenarios where workflow is not associated with the event.
+            try {
+                workflowService.addRequestEntityRelationships(uuid, new Entity[]{new Entity(fullyQualifiedName,
+                        "USER")});
+
+            } catch (InternalWorkflowException e) {
+                //Ignore exception which occurs at DB level since no workflows associated with event
+                log.info("No workflow associated with the operation.");
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/DeleteUserWFRequestHandler.java]: Exception is not logged or thrown

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/bean/Entity.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.workflow.mgt.bean;
+
```
Comment (kasungayan) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/bean/Entity.java]: No comments added to either Class or method definitions. Add some meaningful comments explaining the semantics of this particular class and its methods.

Comment (cdwijayarathna) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/bean/Entity.java]: Thanks, Fixed with https://github.com/wso2/carbon-identity/pull/1675

**Pull request merged**